Emulator/BF.File.Emulator/BfEmulatorApi.cs

```diff
@@ -0,0 +1,99 @@
+﻿using BF.File.Emulator.Interfaces;
+using BF.File.Emulator.Interfaces.Structures.IO;
+using FileEmulationFramework.Interfaces;
+using FileEmulationFramework.Interfaces.Reference;
+using FileEmulationFramework.Lib.Memory;
+using FileEmulationFramework.Lib.Utilities;
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace BF.File.Emulator
+{
+    public class BfEmulatorApi : IBfEmulator
+    {
+        private readonly IEmulationFramework _framework;
+        private readonly BfEmulator _bfEmulator;
+        private readonly Logger _logger;
+
+        public BfEmulatorApi(IEmulationFramework framework, BfEmulator bfEmulator, Logger logger)
+        {
+            //Debugger.Launch();
+            _framework = framework;
+            _bfEmulator = bfEmulator;
+            _logger = logger;
+        }
+
+        RouteFileTuple[] IBfEmulator.GetEmulatorInput()
+        {
+            // Map input to API.
+            var input = _bfEmulator.GetInput();
+            var result = GC.AllocateUninitializedArray<RouteFileTuple>(input.Count);
+            for (int x = 0; x < result.Length; x++)
+            {
+                var original = input[x];
+                result[x] = new RouteFileTuple()
+                {
+                    Route = original.Route.FullPath,
+                    File = original.File
+                };
+            }
+
+            return result;
+        }
+
+        public void InvalidateFile(string bfPath)
+        {
+            _bfEmulator.UnregisterFile(bfPath);
+            _framework.UnregisterVirtualFile(bfPath, false);
+        }
+
+        public void RegisterBf(string sourcePath, string destinationPath)
+        {
+            var handle = Native.CreateFileW(sourcePath, FileAccess.Read, FileShare.ReadWrite, IntPtr.Zero, FileMode.Open, FileAttributes.Normal, IntPtr.Zero);
+            if (handle == new IntPtr(-1))
+            {
+                _logger.Error("[BfEmulatorApi] RegisterBf: Failed to open bf file with Win32 Error: {0}, Path {1}", Marshal.GetLastWin32Error(), sourcePath);
+                return;
+            }
+
+            Native.SetFilePointerEx(handle, 0, IntPtr.Zero, 0);
+
+            var manager = new MemoryManager(65536);
```

Comment:
Purpose of `MemoryManager` is to avoid address space exhaustion in 32-bit programs when having to keep large files in memory.

I believe `.bf` files cap at around 200KiB from what I heard (and many are less than 32KiB); in which case using `MemoryManager` here is not needed (it would be wasteful for <64KiB files in fact). Standard `MemoryStream` is sufficient.

Reply:
I've changed this in 63c2767 so it will now use a `MemoryManagerStream` only for files > 64KiB as when combined with code from multiple mods some bf files do end up being in the hundreds of KiB.

I also added a note to the docs in 32da2d3 about `MemoryManagerStream` with very small files since that's what originally lead me to use it for everything.